File: 백준/Gold/7579. 앱/앱.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main{
    static int N, M, ret = Integer.MAX_VALUE;
    static int[] app;
    static int[] cost;
    static int[][] dp = new int[104][10004]; // 1 ≤ N ≤ 100, 0 ≤ c1, ..., cN ≤ 100
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        app = new int[N + 1];
        cost = new int[N + 1];

        st = new StringTokenizer(br.readLine());
        for(int i = 1; i <= N; i++){
            app[i] = Integer.parseInt(st.nextToken());
        }

        st = new StringTokenizer(br.readLine());
        for(int i = 1; i <= N; i++){
            cost[i] = Integer.parseInt(st.nextToken());
        }

        for(int i = 1; i <= N; i++){

            for(int j = 0; j <= 10000; j++){

                if(cost[i] > j){ // 앱 종료 비용 > 비용
                    dp[i][j] = dp[i - 1][j]; // 갱신 X
                }else{ // 앱 종료 비용 <= 비용
                    dp[i][j] = Math.max(dp[i - 1][j], dp[i - 1][j - cost[i]] + app[i]); // max(갱신 X, 갱신 O)
                    if(dp[i][j] >= M){
                        ret = Math.min(ret, j);
                    }
                }
            }
        }

        System.out.println(ret);
    }
}

File: 백준/Gold/7579. 앱/앱.cc
#include <bits/stdc++.h>
using namespace std;
int n, k; 
int m[104], c[104];
int dp[104][10004]; // dp[i번째][j비용] = 최대메모리크기
int ret = 10000004;
int main(){
    ios_base::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    
    cin >> n >> k;

    for(int i = 1; i <= n; i++){
        cin >> m[i];
    }

    for(int i = 1; i <= n; i++){
        cin >> c[i];
    }

    for(int i = 1; i <= n; i++){
        for(int j = 0; j <= 10000; j++){
            if(c[i] <= j){
                dp[i][j] = max(dp[i - 1][j - c[i]] + m[i], dp[i - 1][j]);

                if(dp[i][j] >= k){
                    ret = min(ret, j);
                }      
            }else dp[i][j] = dp[i - 1][j];
        }
    }

    cout << ret << "\n";

    return 0;
}